pkg/apply/task/apply_task.go

```diff
@@ -162,26 +169,74 @@ func (a *ApplyTask) Start(taskContext *taskrunner.TaskContext) {
 			}
 			if err != nil {
 				if klog.V(4).Enabled() {
-					klog.Errorf("error applying (%s/%s) %s", info.Namespace, info.Name, err)
+					klog.Infof("error applying (%s/%s) %s", info.Namespace, info.Name, err)
 				}
 				taskContext.SendEvent(a.createApplyFailedEvent(
 					id,
 					applyerror.NewApplyRunError(err),
 				))
 				taskContext.InventoryManager().AddFailedApply(id)
-			} else if info.Object != nil {
-				acc, err := meta.Accessor(info.Object)
-				if err == nil {
-					uid := acc.GetUID()
-					gen := acc.GetGeneration()
-					taskContext.InventoryManager().AddSuccessfulApply(id, uid, gen)
+				continue
+			}
+
+			if info.Object == nil {
+				klog.V(5).Infof("apply response missing object (%s/%s)", info.Namespace, info.Name)
+				continue
+			}
+
+			u, ok := info.Object.(*unstructured.Unstructured)
+			if !ok {
+				uMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(info.Object)
+				if err != nil {
+					klog.V(5).Infof("error converting apply response to unstructured (%s/%s): %v", info.Namespace, info.Name, err)
+					continue
 				}
+				u = &unstructured.Unstructured{Object: uMap}
+			}
+
+			klog.V(5).Infof("apply successful (%s/%s)", info.Namespace, info.Name)
+			taskContext.InventoryManager().AddSuccessfulApply(id, u.GetUID(), u.GetGeneration())
+
+			// Update status to shorten wait time with large object sets
+			err = a.updateStatus(ctx, taskContext, u)
+			if err != nil {
+				klog.V(5).Infof("error updating status (%s/%s): %v", info.Namespace, info.Name, err)
 			}
 		}
 		a.sendTaskResult(taskContext)
 	}()
 }
 
+func (a *ApplyTask) updateStatus(ctx context.Context, taskContext *taskrunner.TaskContext, obj *unstructured.Unstructured) error {
+	id := object.UnstructuredToObjMetadata(obj)
+	clusterReaderFactory := engine.ClusterReaderFactoryFunc(clusterreader.NewCachingClusterReader)
+	clusterReader, err := clusterReaderFactory.New(a.Client, a.Mapper, object.ObjMetadataSet{id})
+	if err != nil {
+		return fmt.Errorf("error creating ClusterReader: %w", err)
+	}
+
+	statusReaders, defaultStatusReader := polling.DefaultStatusReaders(a.Mapper)
+	statusReaders = append(statusReaders, defaultStatusReader)
+	for _, statusReader := range statusReaders {
+		if !statusReader.Supports(id.GroupKind) {
+			continue
+		}
```

Comment:
I think you can pass in a map instead of  a list of engine.StatusReader. Then it avoids triggeirng `DefaultStatusReaders` and avoids the for loop in line 220.

Reply:
The impl here is a hack. I'm gonna refactor this code into kstatus somehow. I think we need to make a MetaStatusReader impl that loops over a list of StatusReaders. Also, I'm not sure why the list of StatusReaders doesn't just include the default StatusReader at the end, but maybe it's because the author wanted to allow users to add new ones before the default one but after the others... I think I can get around that by just having the entire list be a single injected MetaStatusReader. Then I can just drop the list here and use the same MetaStatusReader as is injected into the StatusPoller.

Reply:
Sounds good